Escape backward slashes in synced folder id

legal90 opened the pull request:
That fixes invalid share names in Windows guests, where backward slash
character is used as a path separator.

_Details:_
Currently, if we define the synced folder this way (for Windows guest):
```ruby
config.vm.synced_folder "../", 'C:\Users\vagrant\my_share'
```
then the share id will be this string: `'C_\Users\vagrant\my_share'`
In the guest it will be exposed as `\\psf\C_Usersvagrantmy_share` - Parallels Desktop tries to replace backslashes with some special character by itself. Meanwhile the Vagrant expects that the 
share is available at `\\psf\C_Usersvagrantmy_share`, so it creates a broken symlink.

This Pull-Request adds `\` to the regex, so this char will be replaced with `_`.
The result share name will be `'C__Users_vagrant_my_share'` and it will be available at `\\psf\C__Users_vagrant_my_share` without any issues.

The similar PR has been sent to VirtualBox provider: https://github.com/mitchellh/vagrant/pull/8433

romankulikov commented:
Two points from my side:
0. Why does Vagrant expects path `\\psf\C_Usersvagrantmy_share` (without special Unicode symbols)? May be this the issue in Vagrant?
1. Parallels Desktop also remaps other symbols: `"`, `*`, `:`, `<`, `>`, `?`, `|` – may they should be handled as well?

legal90 commented:
> * Parallels Desktop also remaps other symbols: `"`, `*`, `:`, `<`, `>`, `?`, `|` – may they should be handled as well?

Yes, good point! Currently we handle only `:` and `/` here. I'll add other characters too.

> * Why does Vagrant expects path `\\psf\C_Usersvagrantmy_share` (without special Unicode symbols)? May be this the issue in Vagrant?
At the early stage, before booting the VM, our provider configures it by setting the list of shared folders. It calls driver method `#share_folders`, which actually does`prlctl set <vm_uuid> --shf-host-add <share_name> --path <host_path>`. For _<share_name>_ it uses the value of `:name` param if it is set explicitly in Vagrantfile. Otherwise [the guest's path will be used](https://github.com/mitchellh/vagrant/blob/6d3ab39b27d7cc4036c01b17d703bdd5dbe1b8ee/plugins/kernel_v2/config/vm.rb#L217-L221). 

After the VM boot our provider determines the share name from the output of drivers method `#read_shared_folder`, e.q. from the output of `prlctl list -i <vm_uuid> --json`:
https://github.com/Parallels/vagrant-parallels/blob/4025015c9dd1b086317e14895188e42c9f77ba9e/lib/vagrant-parallels/synced_folder.rb#L55-L55

```
In our example `prlctl` returns the following:
...
			"C_\Users\vagrant\my_share": {
				"enabled": true,
				"path": "\/Users\/legal\/Documents",
				"mode": "rw"
			},
...
```
Since backslashes are not escaped in the output (which looks like a bug of prlctl, BTW), the key is decoded exactly as a string `'C_Usersvagrantmy_share'`.
But anyway - since Parallels Desktop has remapped the share name in Windows guest internally, then Vagrant can't handle it anymore. So here we should avoid all cases when the actual share name is different from the expected one.

legal90 commented:
I've updated PR by adding more characters to the escaping regex.
Also I've updated my previous comment: both of setting and mounting of shared folders happens subsequently, after the VM boot.

romankulikov commented:
Ok. Thanks for detailed explanation!
